galsim/phase_psf.py

```diff
@@ -0,0 +1,845 @@
+    @param alpha         Square root of fraction of phase that is "remembered" between time_steps
+                         (i.e., alpha**2 is the fraction remembered). The fraction sqrt(1-alpha**2)
+                         is then the amount of turbulence freshly generated in each step.  Setting
+                         alpha=1.0 results in a frozen-flow atmosphere.  Note that computing PSFs
+                         from frozen-flow atmospheres may be significantly faster than computing
+                         PSFs with non-frozen-flow atmospheres.  [Default: 1.0]
+    @param rng           Random number generator as a galsim.BaseDeviate().  If None, then use the
+                         clock time or system entropy to seed a new generator.  [Default: None]
+
+    Relevant SPIE paper:
+    "Remembrance of phases past: An autoregressive method for generating realistic atmospheres in
+    simulations"
+    Srikar Srinath, Univ. of California, Santa Cruz;
+    Lisa A. Poyneer, Lawrence Livermore National Lab.;
+    Alexander R. Rudy, UCSC; S. Mark Ammons, LLNL
+    Published in Proceedings Volume 9148: Adaptive Optics Systems IV
+    September 2014
+    """
+    def __init__(self, screen_size, screen_scale=None, altitude=0.0, time_step=0.03,
+                 r0_500=0.2, L0=25.0, vx=0.0, vy=0.0, alpha=1.0, rng=None):
+
+        if screen_scale is None:
+            screen_scale = 0.5 * r0_500
+        super(AtmosphericScreen, self).__init__(screen_size, screen_scale, altitude)
+
+        self.time_step = time_step
+        self.r0_500 = r0_500
+        self.L0 = L0
+        self.vx = vx
+        self.vy = vy
+        self.alpha = alpha
+
+        if rng is None:
+            rng = galsim.BaseDeviate()
+        self.orig_rng = rng
+
+        self._init_psi()
+
+        self.reset()
+
+        # Free some RAM for frozen-flow screen
+        if self.alpha == 1.0:
+            del self.psi, self.screen
+
+    def __str__(self):
+        return "galsim.AtmosphericScreen(altitude=%s)" % self.altitude
+
+    def __repr__(self):
+        outstr = ("galsim.AtmosphericScreen(%r, %r, altitude=%r, time_step=%r, " +
+                  "r0_500=%r, L0=%r, vx=%r, vy=%r, alpha=%r, rng=%r)")
+        return outstr % (self.screen_size, self.screen_scale, self.altitude, self.time_step,
+                         self.r0_500, self.L0, self.vx, self.vy, self.alpha, self.rng)
+
+    def __eq__(self, other):
+        sL0 = self.L0 if self.L0 is not None else np.inf
+        oL0 = other.L0 if other.L0 is not None else np.inf
+        return (self.screen_size == other.screen_size and
+                self.screen_scale == other.screen_scale and
+                self.altitude == other.altitude and
+                self.r0_500 == other.r0_500 and
+                sL0 == oL0 and
+                self.vx == other.vx and
+                self.vy == other.vy and
+                self.alpha == other.alpha and
+                self.rng == other.rng)
```

Comment:
Similarly, I don't think this is right.  It doesn't check origin or other aspects of the mutable state:
```
>>> rng = galsim.BaseDeviate()
>>> atm1 = galsim.AtmosphericScreen(10, rng=rng, vx=10, vy=20)
>>> atm2 = galsim.AtmosphericScreen(10, rng=rng, vx=10, vy=20)
>>> atm1 == atm2
True
>>> atm1.advance()
>>> atm1 == atm2
True
>>> atm1.origin
array([-0.3, -0.6])
>>> atm2.origin
array([ 0.,  0.])
```